cargo-miri/src/phases.rs

```diff
@@ -93,8 +93,30 @@ pub fn phase_cargo_miri(mut args: impl Iterator<Item = String>) {
     let target = get_arg_flag_value("--target");
     let target = target.as_ref().unwrap_or(host);
 
-    // We always setup.
-    let miri_sysroot = setup(&subcommand, target, &rustc_version, verbose);
+    let has_build_std = is_build_std_set();
+    if has_build_std {
+        if env::var_os("MIRI_SYSROOT").is_some() {
+            println!("WARNING: `MIRI_SYSROOT` has no effect when Cargo has `build-std` enabled.");
+        }
+        if env::var_os("MIRI_NO_STD").is_some() {
+            println!("WARNING: `MIRI_NO_STD` has no effect when Cargo has `build-std` enabled.");
+        }
+    }
+    let cargo_cmd = match subcommand {
```

Comment:
Why did you move the `cargo_cmd` logic up? This is confusing now, since the `match` has too many jobs.

Also we want to call `setup` before any of the other logic here. I think this should be more like
```rust
let miri_sysroot = if has_build_std {
  Some(setup(...))
} else { None };
```